KEP-3633: graduate matchLabelKeys/mismatchLabelKeys to beta (#4450)
kubernetes/enhancements: keps/sig-scheduling/3633-matchlabelkeys-to-podaffinity/README.md

Reviewer: Can you update the test section with links to the integration tests added for this feature?

  - `matchLabelKeys`/`mismatchLabelKeys` with the feature gate enabled/disabled.

**Filter**
- `k8s.io/kubernetes/test/integration/scheduler/filters/filters_test.go`: https://storage.googleapis.com/k8s-triage/index.html?test=TestPodTopologySpreadFilter
- [`k8s.io/kubernetes/test/integration/scheduler/filters/filters_test.go`](https://github.com/kubernetes/kubernetes/blob/3a4c35cc89c0ce132f8f5962ce4b9a48fae77873/test/integration/scheduler/filters/filters_test.go#L807-L1069): https://storage.googleapis.com/k8s-triage/index.html?test=TestPodTopologySpreadFilter

**Score**
- `k8s.io/kubernetes/test/integration/scheduler/scoring/priorities_test.go`: https://storage.googleapis.com/k8s-triage/index.html?test=TestPodTopologySpreadScoring

Also, we should make sure this feature brings no significant performance degradation in both Filter and Score.

> - `k8s.io/kubernetes/test/integration/scheduler_perf/scheduler_perf_test.go`: https://storage.googleapis.com/k8s-triage/index.html?test=BenchmarkPerfScheduling

Author: After we change the way to implement this feature, this feature no longer has an impact on the scheduling latency since this feature just modifies `labelSelector` and doesn't change anything in the scheduling flow. So, we don't implement any new test case in scheduler_perf for this feature.

Reviewer: Just to clarify for the PRR: this is already implemented with no impact to scheduling.

will rollout across nodes.
-->

> It shouldn't impact already running workloads. It's an opt-in feature,

Reviewer 2: Can't comment on unchanged lines, so commenting here:

> Are there any tests for feature enablement/disablement?

Were those added?
In particular the comment from the template is relevant here.

Author: No... I'll add the one within this release, fortunately it's not too late like minDomains 😓

Author: @wojtek-t By the way, do we have to have a manual test (`Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?`) then? Is there any expectation difference between this automated enablement/disablement test and the manual test?

Reviewer: The manual test should be performed as well. You can put the intent of the test in the KEP now and send an update later indicating that you executed it.

and users need to set `matchLabelKeys` or `mismatchLabelKeys` field in PodAffinity or PodAntiAffinity to use this feature. 

When this feature is disabled by the feature flag, 
the already created Pod's `matchLabelKeys`/`mismatchLabelKeys` is kept and the `labelSelector` is not modified back.
But, the newly created Pod's `matchLabelKeys` or `mismatchLabelKeys` field is ignored and silently dropped.

###### What specific metrics should inform a rollback?

<!--
What signals should users be paying attention to when the feature is young
that might indicate a serious problem?
-->

- A spike on metric `schedule_attempts_total{result="error|unschedulable"}` when pods using this feature are added.

No need to check latency of the scheduler because the scheduler doesn't get changed at all for this feature.

> The only possibility is the bug in the Pod creation process in kube-apiserver and it results in some unintended scheduling.

Reviewer 2: Is this really true? The computations on scheduler side might actually be slightly more expensive no? [e.g. a bit larger selectors to compute against pods...]
@alculquicondor - thoughts?

Reviewer: Maybe... In some scenarios the users might be migrating from manually setting labels to using the new automated way.

Author: > The computations on scheduler side might actually be slightly more expensive no?
> a bit larger selectors to compute against pods

This is correct though, is it realistic that worth being added here?: the labelselector is added by matchLabelKeys → the time the scheduler takes to calculate the labelselector becomes too big. 
If it happens, it's likely not the problem of this feature, but the problem of the entire PodAffinity that labelSelector calculation is somehow buggy.

Reviewer 2: Maybe it's not worth adding, but also please don't mislead the reader that "latency of scheduler doesn't get changed" - it is implicitly changed by using its features more.

Author: I added the mention of the scheduler's latency, but with a note that we could care less about that.

will rollout across nodes.
-->

It shouldn't impact already running workloads. It's an opt-in feature,

> and users need to set `matchLabelKeys` or `mismatchLabelKeys` field in PodAffinity or PodAntiAffinity to use this feature. 

Reviewer: Furthermore, the node selectors are not modified back by the apiserver for already created Pods.

are missing a bunch of machinery and tooling and can't do that now.
-->

We'll test it via this scenario:

1. start kube-apiserver with this feature disabled. 
2. create one Pod with `matchLabelKeys` in PodAffinity. 
3. No change should be observed in `labelSelector` in PodAffinity because the feature is disabled.
4. restart kube-apiserver with this feature enabled. (**First enablement**)
5. No change in Pod created at (2).
6. create one Pod with `matchLabelKeys` in PodAffinity.
7. `labelSelector` in PodAffinity should be changed based on `matchLabelKeys` and label value in the Pod because the feature is enabled.
8. restart kube-apiserver with this feature disabled. (**First disablement**)
9. No change in Pods created before.
10. create one Pod with `matchLabelKeys` in PodAffinity. 
11. No change should be observed in `labelSelector` in PodAffinity because the feature is disabled.
12. restart kube-apiserver with this feature enabled. (**Second enablement**)
13. No change in Pods created before.
14. create one Pod with `matchLabelKeys` in PodAffinity.

> 15. `labelSelector` in PodAffinity should be changed based on `matchLabelKeys` and label value in the Pod because the feature is enabled.

Reviewer 2: This scenarios is fine - please report to the KEP back after running it [doesn't block the KEP merge, but please block the actualy graduation on it]

[existing SLIs/SLOs]: https://git.k8s.io/community/sig-scalability/slos/slos.md#kubernetes-slisslos
-->

Yes. there is an additional work: kube-apiserver uses the keys in `matchLabelKeys` or `mismatchLabelKeys` to look up label values from the pod, 
and change `labelSelector` according to them.

> Maybe result in a very small impact in the latency of pod creation request in kube-apiserver. 

Reviewer: ```suggestion
The impact in the latency of pod creation request in kube-apiserver should be negligible. 
```
